src/graph/optimizer/OptRule.cpp

```diff
@@ -59,26 +62,68 @@ StatusOr<MatchedResult> Pattern::match(const OptGroupNode *groupNode) const {
   for (size_t i = 0; i < dependencies_.size(); ++i) {
     auto group = groupNode->dependencies()[i];
     const auto &pattern = dependencies_[i];
-    auto status = pattern.match(group);
+    auto status = pattern.match(group, boundary);
     NG_RETURN_IF_ERROR(status);
     result.dependencies.emplace_back(std::move(status).value());
   }
   return result;
 }
 
-StatusOr<MatchedResult> Pattern::match(const OptGroup *group) const {
+StatusOr<MatchedResult> Pattern::match(const OptGroup *group,
+                                       std::vector<OptGroup *> &boundary) const {
   for (auto node : group->groupNodes()) {
-    auto status = match(node);
+    auto status = match(node, boundary);
     if (status.ok()) {
       return status;
     }
   }
   return Status::Error();
 }
 
-StatusOr<MatchedResult> OptRule::match(OptContext *ctx, const OptGroupNode *groupNode) const {
+bool OptRule::TransformResult::checkDataFlow(const std::vector<OptGroup *> &boundary) {
+  return std::all_of(
+      newGroupNodes.begin(), newGroupNodes.end(), [&boundary](const OptGroupNode *groupNode) {
+        return checkDataFlow(groupNode, boundary);
+      });
+}
+
+/*static*/ bool OptRule::TransformResult::checkDataFlow(const OptGroupNode *groupNode,
+                                                        const std::vector<OptGroup *> &boundary) {
+  const auto &deps = groupNode->dependencies();
+  // reach the boundary
+  if (std::all_of(deps.begin(), deps.end(), [&boundary](OptGroup *dep) {
+        return std::find(boundary.begin(), boundary.end(), dep) != boundary.end();
+      })) {
+    return true;
+  }
+  const auto *group = groupNode->group();
+  if (std::find(boundary.begin(), boundary.end(), group) != boundary.end()) {
+    return true;
+  }
+  // Check dataflow
+  const auto *node = groupNode->node();
+  if (node->inputVars().size() == deps.size()) {
+    // Don't check when count of dependencies is different from count of input variables
+    for (std::size_t i = 0; i < deps.size(); i++) {
+      const OptGroup *dep = deps[i];
+      if (node->inputVar(i) != dep->outputVar()) {
+        return false;
+      }
```

Comment:
Maybe we could do more things than this input variable validation, such as check whether these new generated OptGroupNodes reference other nodes outside since we guarantee that all plan nodes of the new subplan couldn't access other plan nodes except the intput/output nodes.

Reply:
I this has been done in current implementation. All input variables of plan nodes in current transform result had been checed.

Reply:
Fixed.